### ISSUE TITLE
🦠 Model Request: HIA

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#425]
### Model Title

HIA (TDC dataset)

### Publication

Hello @pauline-banye !

As part of your Outreachy contribution, we have assigned you the dataset "HIA" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by pauline-banye** [comment]
Hi @GemmaTuron, my apologies for the delayed response. I have been under the weather for a couple of days but I'm getting much better. My work on this has been sporadic but I would like to provide an update. 

**OVERVIEW**
Theoretical models for the prediction of absorption, distribution, metabolism and excretion (ADME) properties play extremely important roles in support of the drug development process. 

I was tasked with building a binary classification  model for  HIA. Human intestinal absorption is the ability of the human gastrointestinal system to absorb orally administered drugs into the bloodstream. HIA is a vital factor influencing the transportation of drugs to the intended targets in the body. 

From my understanding, the primary goal of predicting HIA is to measure human oral absorption, which is a determining factor in early drug discovery. Due to the complexity of drug absorption, it's challenging to analyse and build useful statistical models for diverse sets of pharmaceuticals.  


**TASKS**
- [x] STEP 1: Install Therapeutics Data Commons
- [x] STEP 2: Load your data from TDC using the specific assay assigned to you
- [x] STEP 3: Get the train/validation/test splits from TDC
    -  This was carried out by splitting the dataset into 3 parts 
        - The training set which is used to teach the model. 
        - The validation set which is utilised for improving the model.
        - The test set for evaluating the model's performance.
![lzy](https://user-images.githubusercontent.com/63536707/198160461-1d50cb56-56ed-452c-9daf-c1f935e960a8.png)
- [x] STEP 4: Save the datasets in your Google Drive as .csv files using the pandas package

**Comment by pauline-banye** [comment]
- [x] STEP 5: How many molecules do you have in the dataset? How many active and inactives?
   - The HIA dataset from TDC comprises of a total of 578 molecules, 500 of which are active and 78 inactive.
      - 404 molecules are present in the train set, with 351 active and 53 inactive molecules.
      - 58 molecules are present in the valid set, consisting of 52 active and 6 inactive molecules
      - 116 molecules are present in the test set of which 97 are active and 19 are inactive molecules,
- [x] STEP 6: understand your data - use the matplotlib package to plot the data outcomes
     Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations in Python. I utilised this library to create a bar graph and pie chart for visual interpretation of the resulting data.
     - Using a bar graph
     ![lzy](https://user-images.githubusercontent.com/63536707/198164306-7a97bfd2-33e0-42bf-bd38-c2e616ade622.png)
     - Using a piechart
     ![lzy](https://user-images.githubusercontent.com/63536707/198165924-3f9f3df5-b68c-480f-8861-8e2379e630de.png)
    From the results of our binary classifier, we receive two outputs:
    - Probability of 0: the probability that a molecule is inactive in the given assay
    - Probability of 1: the probability that a molecule is active in the given assay
    - From the charts, it is visible that in this train set, the amount of active molecules far outweighs the inactive molecules resulting in an imbalanced set

**Comment by pauline-banye** [comment]
> * [x]  STEP 5: How many molecules do you have in the dataset? How many active and inactives?
>   
>   * The HIA dataset from TDC comprises of a total of 578 molecules, 500 of which are active and 78 inactive.
>     
>     * 404 molecules are present in the train set, with 351 active and 53 inactive molecules.
>     * 58 molecules are present in the valid set, consisting of 52 active and 6 inactive molecules
>     * 116 molecules are present in the test set of which 97 are active and 19 are inactive molecules,
> * [x]  STEP 6: understand your data - use the matplotlib package to plot the data outcomes
>   Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations in Python. I utilised this library to create a bar graph and pie chart for visual interpretation of the resulting data.
>   
>   * Using a bar graph
>     ![lzy](https://user-images.githubusercontent.com/63536707/198164306-7a97bfd2-33e0-42bf-bd38-c2e616ade622.png)
>   * Using a piechart
>     ![lzy](https://user-images.githubusercontent.com/63536707/198165924-3f9f3df5-b68c-480f-8861-8e2379e630de.png)
>     From the results of our binary classifier, we receive two outputs:
>   * Probability of 0: the probability that a molecule is inactive in the given assay
>   * Probability of 1: the probability that a molecule is active in the given assay
>   * I observed that a vast majority of the molecules in the HIA dataset are in an active state
>   * From the charts, it is visible that in this train set, the amount of active molecules far outweighs the inactive molecules resulting in an imbalanced set

Hi @GemmaTuron , I have a question regarding the results of the split. I could tell that the data was imbalanced and I remember reading that working with an imbalanced data can reduce the efficacy of the model and there are methods such as resampling, oversampling, precision to name a few. 
- Could you please provide some more clarity on the effects of imbalanced data?
- Is it required in this case to try and _balance_ the data using any of such methods?

**Comment by pauline-banye** [comment]
- [x] STEP 7: check your molecules - draw an active molecule and an inactive molecule (use the RDKIT package)
   The rdkit library is a Python library that allows us to handle chemical structures and the calculation of their molecular properties. I used the rdkit library to generate visual representations of the first 3 active and inactive molecules.

    **_active molecules_**
    ![lzy](https://user-images.githubusercontent.com/63536707/198170276-ee6db746-4698-4afc-8c69-2d7da706bdfd.png)

    **_inactive molecules_**
    ![lzy](https://user-images.githubusercontent.com/63536707/198170671-ae20a4bf-be8b-4aff-a97e-3afed4ae9be3.png)

    - [x] STEP 8: Create your lists of Y_Train, Y_val, Y_test.
    - [x] STEP 9: Install Lazy-Qsar from GitHub and import the Morgan Binary Classifier class.
        - I cloned the lazyqsar repo
        - installed lazyqsar with pip -e for developer mode
        - imported the Morgan Binary Classifier class to convert the smiles to vector which the computer can understand
    - [x] STEP 10: Train a Binary Classifier
    - [x] STEP 11: Save the model in joblib format

**Comment by pauline-banye** [comment]
- [x] STEP 9: Install Lazy-Qsar from GitHub and import the Morgan Binary Classifier class.
         LazyQsar is a library that enables us to build models quickly.  It is extremely convenient as it  converts our smiles into vectors automatically,  which can be interpreted by the computer system.
        - I cloned the lazyqsar repo
        - installed lazyqsar with pip -e for developer mode
        - imported the Morgan Binary Classifier class to convert the smiles to vector which the computer can understand
        
    - [x] STEP 10: Train a Binary Classifier
        - Used a process called FIT where we input the smiles and results.
        - The LazyQSAR package automatically converts the model to morgan fingerprints
        
    - [x] STEP 11: Save the model in joblib format

**Comment by pauline-banye** [comment]
- [x] STEP 9: Install Lazy-Qsar from GitHub and import the Morgan Binary Classifier class.  
   LazyQsar is a library that enables us to build models quickly.  It is extremely convenient as it  converts our smiles into vectors automatically,  which can be interpreted by the computer system. 

- - I cloned the lazyqsar repo
- - installed lazyqsar with pip -e for developer mode
- - imported the Morgan Binary Classifier class to convert the smiles to vector which the computer can understand
- 

- [x] STEP 10: Train a Binary Classifier

- -  Used a process called FIT where we input the smiles and results.
- -  The LazyQSAR package automatically converts the model to morgan fingerprints
-         

- [x] STEP 11: Save the model in joblib format

**Comment by GemmaTuron** [comment]
Hi @pauline-banye !

very good job, please also add your comments on model performance before we close this contribution!

**Comment by pauline-banye** [comment]
Once the model was trained, I evaluated the model performance by predicting the results using the validation and test sets. The list of smiles data and the result is passed to the model and compared to the actual value. 

The outcome of this classification process is a probability of 0 or 1, which can be either a true or false positive or a true or false negative.
- An outcome is a true positive when the model correctly predicts that a molecule belongs to the positive class. 
- A true negative outcome is one where the model correctly predicts the molecule as belonging to a negative class. 
- However, when the model incorrectly predicts a molecule as a positive class, it is termed a false positive.
- Similarly, a false negative is an outcome where the model incorrectly predicts the molecule as a negative class. 

Note that these values are dependent on the threshold set for the classification. The default threshold is 0.5 but if the threshold is lowered to 0.4 for example, molecules with a probability of 0.4 and were previously classified as negative would be classified as positive.

The metrics used for evaluating model performance on the validation and test sets are:
1. AUROC value
 The Area Under the Curve (AUC) measures the ability of a classifier to distinguish between classes. The higher the AUROC value, the better the performance of the model at identifying distinct classes. An AUROC value of 0.7 - 0.8 is acceptable, 0.8 - 0.9 is excellent, and more than 0.9 is outstanding.
   - validation set - AUROC 0.9775641025641025
   - test set - AUROC 0.9706999457406403
 For this model, the AUROC values were outstanding:
 
2. ROC Curve
A receiver operating characteristic curve, or ROC curve, is a graphical plot that illustrates the performance of a binary classification model.  The ROC curve is plotted with TruePositiveRate(TPR) against the FalsePosotiveRate(FPR) where TPR is on the y-axis and FPR is on the x-axis. Classifiers that give curves closer to the top-left corner indicate a better performance

_Validation set ROC Curve_
![roc1](https://user-images.githubusercontent.com/63536707/198618340-eb23c3df-9e45-4354-ad89-ab754c0277a6.png)

_Test set ROC Curve_
![roc1](https://user-images.githubusercontent.com/63536707/198620373-9a5c8e44-ef86-4a2c-923a-4245378c66be.png)

This model exhibits an extremely high performance as it's plot is almost vertical along the x axis.

3. Contingency Table
4. Precision score
5. Recall score
6. Accuracy

**Comment by pauline-banye** [comment]
1. AUROC value
 The Area Under the Curve (AUC) measures the ability of a classifier to distinguish between classes. The higher the AUROC value, the better the performance of the model at identifying distinct classes. An AUROC value of 0.7 - 0.8 is acceptable, 0.8 - 0.9 is excellent, and more than 0.9 is outstanding.

   - validation set - AUROC 0.9775641025641025
   - test set - AUROC 0.9706999457406403
   
For this model, the AUROC values were outstanding:

**Comment by pauline-banye** [comment]
2. ROC Curve
A receiver operating characteristic curve, or ROC curve, is a graphical plot that illustrates the performance of a binary classification model.  The ROC curve is plotted with TruePositiveRate(TPR) against the FalsePosotiveRate(FPR) where TPR is on the y-axis and FPR is on the x-axis. Classifiers that give curves closer to the top-left corner indicate a better performance

_Validation set ROC Curve_
![roc1](https://user-images.githubusercontent.com/63536707/198618340-eb23c3df-9e45-4354-ad89-ab754c0277a6.png)

_Test set ROC Curve_
![roc1](https://user-images.githubusercontent.com/63536707/198620373-9a5c8e44-ef86-4a2c-923a-4245378c66be.png)

This model exhibits an extremely high performance as it's plot is almost vertical along the x axis.

**Comment by pauline-banye** [comment]
3. Contingency Table
Contingency tables are used to record the number of
molecules assigned to different classes after a test has been performed. It displays a visual representation of the quantity of molecules which are correctly or incorrectly assigned  to each specific class.

_Validation set - Contingency table_
![roc1](https://user-images.githubusercontent.com/63536707/198643996-94ef68f7-477b-4426-be09-6b419125e6cb.png)

From the contingency table above,  we can determine that out of the total 58 molecules in the valid set, 52 were correctly classified as active, 3 were correctly classified as negative, 3 were incorrectly classified as positive and none were false negatives.

_Test  set - Contingency table_
![roc1](https://user-images.githubusercontent.com/63536707/198643144-5996f5be-1721-4dda-a685-d1f35d7ff0c6.png)

From the contingency table,  we can determine that out of the total 58 molecules in the valid set, 97 were correctly classified as active, 13 were correctly classified as negative, 6 were incorrectly classified as positive and none were false negatives.

**Comment by pauline-banye** [comment]
> Hi @pauline-banye !
> 
> very good job, please also add your comments on model performance before we close this contribution!

Thanks so much @GemmaTuron 😊! 
I'm currently working on the few remaining evaluations. I just need to add the precision score, recall score and accuracy, then I should be done!

**Comment by pauline-banye** [comment]
5. Precision score
Precision is one of the indicators of a machine learning model's performance. It outlines the quality of a positive prediction made by the model. It is a measure of the number of positive class predictions that actually belong to the positive class.

It is calculated as the number of true positives divided by the total number of positive predictions (i.e., the number of true positives plus the number of false positives). The best value is 1 and the worst value is 0.

_Validation set - Precision score_
![roc1](https://user-images.githubusercontent.com/63536707/198674833-f09d9dac-a1c9-4bfe-b697-6441519b335e.png)

_Test set - Precision set_
![roc1](https://user-images.githubusercontent.com/63536707/198673296-139e5d6a-14fc-40d1-9351-6e29d3fd966a.png)


The precision scores for the validation set and test sets indicate that approximately 0.95 and 0.94 respectively of the positive identifications are actually correct.


7. Recall score
Recall quantifies the number of positive class predictions made out of all positive examples in the dataset.

_Validation set - Recall score_
![roc1](https://user-images.githubusercontent.com/63536707/198675938-7c50ab7d-34af-4e79-9d11-5becbd3d0fa8.png)


_Test set - Recall score_
![roc1](https://user-images.githubusercontent.com/63536707/198673794-a063d1ba-4a7d-4515-88c8-9573c49bc14d.png)


9. Accuracy

**Comment by pauline-banye** [comment]
7. Recall score
    Recall tries to answer the question - how many positives are we able to identify? It is calculated as the ratio between the numbers of positive molecules which were correctly identified as Positive to the total number of Positive molecules in the set.  

   Recall quantifies the number of positive class predictions made out of all positive examples in the dataset. The higher the recall value, the more positive samples detected. The highest recall value for a good classifier should be 1 with 0 being the lowest value.

    _Validation set - Recall score_
    ![roc1](https://user-images.githubusercontent.com/63536707/198675938-7c50ab7d-34af-4e79-9d11-5becbd3d0fa8.png)

    _Test set - Recall score_
    ![roc1](https://user-images.githubusercontent.com/63536707/198673794-a063d1ba-4a7d-4515-88c8-9573c49bc14d.png)

From the images above, we observe that recall scores for both sets is 1 indicating that the classifier is good. Recall value becomes 1 only when the numerator and denominator are equal i.e True Positives = True Positivies + False Negatives, this also means that False Negatives is zero.

**Comment by pauline-banye** [comment]
6. Accuracy
    Accuracy is one of the metrics used for evaluating classification models. It measures the number of correctly predicted molecules out of all the molecules.. It provides an estimate of how often the model classifies a data point correctly. 

    For binary classification, accuracy can also be calculated in terms of positives and negatives as follows:
(TruePositive + TrueNegative) / (TruePositive + TrueNegative + FalseNegative + FalsePositive)

    _Validation set - Accuracy_
    ![roc1](https://user-images.githubusercontent.com/63536707/198691676-f9c3bff3-11fe-491f-b371-f31b13f39095.png)

    _Test set - Accuracy_
    ![roc1](https://user-images.githubusercontent.com/63536707/198696270-df3625cc-51a6-4774-8c97-00975eaa7282.png)

    Our model has an accuracy score of approximately 0.95(95%). This means that 95% of  the molecules were predicted correctly.
    
    In machine learning,  70% accuracy is a great model performance. An accuracy measure of anything between 70%-90% is not only ideal, it's realistic and consistent with industry standards.

**Comment by pauline-banye** [comment]
In summary, 
- I acquired the HIA data from TDC and prepared the data for analysis.
- I created visualisations for the data - 
- bar chart - Matplotlib
- pie chart - Matplotlib
- molecular structure - RDKit
- I  trained the model and used Morgan Binary Classifier to classify the molecules
- I successfully evaluated the model performance on the validation and test sets using the following metrics:
- -AUROC value: approx 0.98 & 0.97 respectively
- -ROC Curve: Both the validation & test sets were excellent 
- -Contingency Table: Very high accuracy
- -Precision score: approx 0.95 & 0.94 respectively
- -Recall score: Value is 1
- -Accuracy: approx 0.95 for both sets
Overall the model performance on both the validation and test sets was excellent.

**Comment by EstherIdabor** [comment]
@pauline-banye The performance of your model is stellar and your analysis is beautiful, well-done.

**Comment by pauline-banye** [comment]
> @pauline-banye The performance of your model is stellar and your analysis is beautiful, well-done.

@EstherIdabor Thanks so much sis 😊

**Comment by GemmaTuron** [comment]
Hello @pauline-banye !

Great job thanks for the effort! I'll mark this as closed, please focus on your final application to Outreachy.

**Comment by pauline-banye** [comment]
> Hello @pauline-banye !
> 
> Great job thanks for the effort! I'll mark this as closed, please focus on your final application to Outreachy.

Thank you so much @GemmaTuron 😃!
